tests/testthat/test-create_iso8601.R

```diff
@@ -0,0 +1,83 @@
+test_that("`create_iso8601()`: individual date components", {
+
+  x <- c("0", "50", "1950", "80", "1980", "2000")
+  y0 <- create_iso8601(x, .format = "y", .check_format = FALSE)
+  y1 <- c(NA, "2050", "1950", "1980", "1980", "2000")
+  expect_identical(y0, y1)
+
+  x <- c("0", "jan", "JAN", "JaN", "1", "01")
+  y0 <- create_iso8601(x, .format = "m", .check_format = FALSE)
+  y1 <- c(NA, "--01", "--01", "--01", NA, "--01")
+  expect_identical(y0, y1)
+
+  x <- c("0", "00", "1", "01", "10", "31")
+  y0 <- create_iso8601(x, .format = "d", .check_format = FALSE)
+  y1 <- c("----00", "----00", "----01", "----01", "----10", "----31")
+  expect_identical(y0, y1)
+
+})
+
+test_that("`create_iso8601()`: dates", {
+
+  y1 <- c("1999-01-01", "2000-01-01", "1999-01-01", "1999-12-31")
+
+  x <- c("19990101", "20000101", "990101", "991231")
+  y0 <- create_iso8601(x, .format = "ymd", .check_format = FALSE)
+  expect_identical(y0, y1)
+
+  x <- c("1999-01-01", "2000-01-01", "99-01-01", "99-12-31")
+  y0 <- create_iso8601(x, .format = "y-m-d", .check_format = FALSE)
+  expect_identical(y0, y1)
+
+  x <- c("1999 01 01", "2000 01 01", "99 01 01", "99 12 31")
+  y0 <- create_iso8601(x, .format = "y m d", .check_format = FALSE)
+  expect_identical(y0, y1)
+
+})
+
+test_that("`create_iso8601()`: times: hours and minutes", {
+
+  y1 <- c("-----T15:20", "-----T00:10", "-----T23:01", "-----T00:00")
+
+  x <- c("1520", "0010", "2301", "0000")
+  y0 <- create_iso8601(x, .format = "HM", .check_format = FALSE)
+  expect_identical(y0, y1)
+
+  x <- c("15:20", "00:10", "23:01", "00:00")
+  y0 <- create_iso8601(x, .format = "H:M", .check_format = FALSE)
+  expect_identical(y0, y1)
+
+  x <- c("15h20", "00h10", "23h01", "00h00")
+  y0 <- create_iso8601(x, .format = "HhM", .check_format = FALSE)
+  expect_identical(y0, y1)
+
+})
+
+test_that("`create_iso8601()`: times: hours, minutes and seconds", {
+
+  x <- c("152000", "001059", "230112.123", "00002.")
+  y0 <- create_iso8601(x, .format = "HMS", .check_format = FALSE)
+  y1 <- c("-----T15:20:00", "-----T00:10:59", "-----T23:01:12.123", "-----T00:00:02")
+  expect_identical(y0, y1)
+
+  x <- c("15:20:00", "00:10:59", "23:01:12.123", "00:00:2.", "5:1:4")
+  y0 <- create_iso8601(x, .format = "H:M:S", .check_format = FALSE)
+  y1 <- c(y1, "-----T05:01:04")
+  expect_identical(y0, y1)
+
+})
+
+
+test_that("`create_iso8601()`: dates and times", {
+
+  dates <- c("1999-01-01", "2000-01-01", "99-01-01", "99-12-31")
+  times <- c("1520", "0010", "2301", "0000")
+  iso8601_dttm <- create_iso8601(dates, times, .format = c("y-m-d", "HM"), .check_format = FALSE)
+  expectation <-
+    c("1999-01-01T15:20",
+      "2000-01-01T00:10",
+      "1999-01-01T23:01",
+      "1999-12-31T00:00")
+  expect_identical(iso8601_dttm, expectation)
+
+})
```

Comment:
Just to document that I tested with the Roche test cases and apart from three cases it is a match. I believe sdtm.oak behaves correctly.

```
# The only diffs I see are
# - "UN UNK 0" that roak make as NA, meanwhile sdtm.oak "----00"
# - "13NOV 2010", "33:10:12" that roak make as "2010-11-13", meanwhile sdtm.oak "2010-11-13T33:10:12
# - "2010.11.12", "33:93" that roak make as "2010-11-12", meanwhile sdtm.oak "-----T33:93"
```

```R
date_vector <- c(
  "12NOV2020", "12 NOV 2020", "12 NOV 2020", "03 NOV 2020",
  "UN DEC 2019", "3 NOV 2020", "NOV 2020", "MAR 2019", "UN APR 2018",
  "UN UNK 2015", "APR 2018", "UK APR 2018", "UN UNK 0", "UN UN 2015",
  "20190212", "20190202", "2019-04-04", "20200505", "33 MAR2020", "13NOV 2010",
  "UN NOV 2020", "UNNOV2020", "03UNK2020", "3NOV2020", "UNUNK2020", "UNUNK2020",
  "23.05.2019", "22-12-2010", "2010.11.12", "UN 1999", "UNK MAR 2020"
)

time_vector <- c(
  NA, NA, "15:15", "3:30", "14:00", NA, NA, NA, "NA", NA, NA,
  NA, NA, NA, NA, "02:20", "1045", "NA", NA, "33:10:12", "3:30", "3:18",
  "3:18", "0318", "0318", NA, "18:20", "1717", "33:93", NA, NA
)

date_formats <- dtc_formats |> 
  dplyr::filter(type %in% "date") |> 
  dplyr::pull(fmt) |> 
  c("dm y", "d.m.y", "y.m.d")

time_formats <- dtc_formats |> 
  dplyr::filter(type %in% "time") |> 
  dplyr::pull(fmt)

# 310K records
# user  system elapsed 
# 7.20    0.34    8.58
system.time({
  observed_output <- create_iso8601(
    date_vector,
    time_vector,
    .format = list(
      date_formats = date_formats, 
      time_formats = time_formats
    ),
    .na = c("UN", "UN UN", "UN UNK", "UNK", "UK"),
    .check_format = FALSE
  )
})

expected_output <- c(
  "2020-11-12", "2020-11-12", "2020-11-12T15:15",
  "2020-11-03T03:30", "2019-12--T14:00", "2020-11-03",
  "2020-11", "2019-03", "2018-04", "2015", "2018-04",
  "2018-04", "----00", "2015", "2019-02-12", "2019-02-02T02:20",
  "2019-04-04T10:45", "2020-05-05", NA, "2010-11-13T33:10:12",
  "2020-11--T03:30", "2020-11--T03:18", "2020---03T03:18",
  "2020-11-03T03:18", "2020----T03:18", "2020", "2019-05-23T18:20",
  "2010-12-22T17:17", "2010-11-12T33:93", "1999", "2020-03"
)
testthat::expect_identical(observed_output, expected_output)
```

Reply:
FYI @rammprasad

Reply:
So in roak, we map `"33:10:12"` to NA/missing because of the 33, right? That's why at a certain point I asked whether to made semantic validation... I guess this is the case here? You find that 33 is nonsensical and invalidate the time, right? If you wish that can be easily introduced in sdtm.oak.

Reply:
In this example: `"UN UNK 0"` have you tried to supply `.na = c("UN", "UNK", "0")`?

Reply:
```
create_iso8601("2010.11.12", "33:93", .format = c("y.m.d", "H:M"), .check_format = FALSE)
[1] "2010-11-12T33:93"
```

We have to be careful here with the dots `"."` or any character that might be interpreted as regex... This can be regarded as a strength or a bug... look at this example:

```
create_iso8601("2010X11Y12", "33:93", .format = c("y.m.d", "H:M"), .check_format = FALSE)
[1] "2010-11-12T33:93"
```
So actually the format `"y.m.d"` means an year, any character, month, any character, day.

Reply:
I also suggest that we mention in the `create_iso8601()`  that format is a regex and in the article provide some bad regexe examples as well to make them aware.